### PR TITLE
Finish implementing TerrainLogic

### PR DESCRIPTION
**Pull request by jonwil**

### CODE REVIEW

File: src/platform/w3dengine/client/w3dterrainlogic.cpp [hunk 1]
```diff
@@ -13,3 +13,246 @@
...
+    }
+
+    m_mapDX = map->Get_X_Extent();
+    m_mapDY = map->Get_Y_Extent();
+    m_boundaries = map->Get_All_Boundaries();
+    m_activeBoundary = 0;
+    int min_z = 255;
+    int max_z = 0;
+
+    for (int i = 0; i < m_mapDY; i++) {
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dterrainlogic.cpp]: y

File: src/platform/w3dengine/client/w3dterrainlogic.cpp [hunk 2]
```diff
@@ -13,3 +13,246 @@
...
+        for (int j = 0; j < m_mapDX; j++) {
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dterrainlogic.cpp]: x

File: src/platform/w3dengine/client/w3dterrainlogic.cpp [hunk 3]
```diff
@@ -13,3 +13,246 @@
...
+            g_theGameClient->Set_Time_Of_Day(g_theWriteableGlobalData->m_timeOfDay);
+        }
+
+        return true;
+    }
+
+    return false;
+}
+
+void W3DTerrainLogic::New_Map(bool b)
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dterrainlogic.cpp:128]: unk

File: src/platform/w3dengine/client/w3dterrainlogic.cpp [hunk 4]
```diff
@@ -13,3 +13,246 @@
...
+
+    return 0.0f;
+}
+
+void W3DTerrainLogic::Get_Extent(Region3D *extent) const
+{
+    extent->lo.x = 0.0f;
+    extent->lo.y = 0.0f;
+
+    if (m_boundaries.size() != 0) {
```
Comment (xezon) [inline src/platform/w3dengine/client/w3dterrainlogic.cpp:198]: This looks suspicious. `m_activeBoundary < m_boundaries.size()` would be the better test.

File: src/game/logic/map/terrainlogic.cpp [hunk 5]
```diff
@@ -813,6 +818,31 @@ float Bridge::Get_Bridge_Height(const Coord3D *loc, Coord3D *n)
...
+TerrainLogic::TerrainLogic() :
+    m_activeBoundary(0),
+    m_waterGridEnabled(false),
+    m_numWaterToUpdate(0),
+    m_waypointListHead(nullptr),
+    m_bridgeListHead(nullptr),
+    m_mapData(nullptr),
+    m_bridgeDamageStatesChanged(false),
+    m_mapDX(0),
+    m_mapDY(0)
```
Comment (xezon) [inline src/game/logic/map/terrainlogic.cpp]: Initialization order does not match declaration order.

File: src/game/logic/map/terrainlogic.cpp [hunk 6]
```diff
@@ -1416,3 +1446,845 @@ void TerrainLogic::Load_Post_Process()
...
+    }
+
+    if (!query) {
+        g_theTerrainVisual->Load(Get_Source_Filename());
+    }
+
+    return true;
+}
+
+void TerrainLogic::New_Map(bool b)
```
Comment (xezon) [inline src/game/logic/map/terrainlogic.cpp:1487]: unk

File: src/game/logic/map/terrainlogic.cpp [hunk 7]
```diff
@@ -1416,3 +1446,845 @@ void TerrainLogic::Load_Post_Process()
...
+{
+    m_waterGridEnabled = enable;
+
+    if (!enable) {
+        g_theTerrainVisual->Enable_Water_Grid(enable);
+        return;
+    }
+
+    int index = -1;
+    for (int i = 0; i < 4; i++) {
```
Comment (xezon) [inline src/game/logic/map/terrainlogic.cpp:1507]: ARRAY_SIZE(g_theWriteableGlobalData->m_vertexWaterAvailableMaps)

File: src/game/logic/map/terrainlogic.cpp [hunk 8]
```diff
@@ -1416,3 +1446,845 @@ void TerrainLogic::Load_Post_Process()
...
+        if (!g_theWriteableGlobalData->m_mapName.Compare_No_Case(
+                g_theWriteableGlobalData->m_vertexWaterAvailableMaps[i].Str())) {
+            index = i;
+        }
+
+        Utf8String str;
+        Utf8String str2;
+        const char *str3 = strrchr(g_theWriteableGlobalData->m_mapName.Str(), '\\');
```
Comment (xezon) [inline src/game/logic/map/terrainlogic.cpp]: fileName

File: src/game/logic/map/terrainlogic.cpp [hunk 9]
```diff
@@ -1416,3 +1446,845 @@ void TerrainLogic::Load_Post_Process()
...
+        if (str3 != nullptr) {
+            str.Set(str3);
+        } else {
+            str = g_theWriteableGlobalData->m_mapName;
+        }
+
+        str3 = strrchr(g_theWriteableGlobalData->m_vertexWaterAvailableMaps[i].Str(), '\\');
+
+        if (str3 != nullptr) {
+            str.Set(str3);
```
Comment (xezon) [inline src/game/logic/map/terrainlogic.cpp]: `str2`, `vertexWaterFileName` I think

File: src/game/logic/map/terrainlogic.cpp [hunk 10]
```diff
@@ -1416,3 +1446,845 @@ void TerrainLogic::Load_Post_Process()
...
+    }
+
+    int index = -1;
+    for (int i = 0; i < 4; i++) {
+        if (!g_theWriteableGlobalData->m_mapName.Compare_No_Case(
+                g_theWriteableGlobalData->m_vertexWaterAvailableMaps[i].Str())) {
+            index = i;
+        }
+
+        Utf8String str;
```
Comment (xezon) [inline src/game/logic/map/terrainlogic.cpp]: mapFileName

File: src/game/logic/map/terrainlogic.cpp [hunk 11]
```diff
@@ -1416,3 +1446,845 @@ void TerrainLogic::Load_Post_Process()
...
+        Utf8String str2;
```
Comment (xezon) [inline src/game/logic/map/terrainlogic.cpp]: vertexWaterFileName

File: src/game/logic/map/terrainlogic.cpp [hunk 12]
```diff
@@ -1416,3 +1446,845 @@ void TerrainLogic::Load_Post_Process()
...
+        "!!!!!! Deformable water won't work because there was no group of vertex water data defined in GameData.INI for "
+        "this map name '%s' !!!!!! (C. Day)",
+        g_theWriteableGlobalData->m_mapName.Str());
+}
+
+Coord3D TerrainLogic::Find_Closest_Edge_Point(const Coord3D *pos) const
+{
+    Region3D region;
+    Get_Extent(&region);
+    float f[4];
```
Comment (xezon) [inline src/game/logic/map/terrainlogic.cpp]: points

File: src/game/logic/map/terrainlogic.cpp [hunk 13]
```diff
@@ -1416,3 +1446,845 @@ void TerrainLogic::Load_Post_Process()
...
+        c.y = region.hi.y;
+    }
+
+    c.z = Get_Ground_Height(c.x, c.y, nullptr);
+    return c;
+}
+
+void Make_Align_To_Normal_Matrix(float angle, const Coord3D &pos, const Coord3D &n, Matrix3D &tm)
+{
+    Coord3D c = n;
```
Comment (xezon) [inline src/game/logic/map/terrainlogic.cpp]: `c` appears to be never modified, so we can just use `n` directly.

Or call it `axis`

File: src/game/logic/map/terrainlogic.cpp [hunk 14]
```diff
@@ -1416,3 +1446,845 @@ void TerrainLogic::Load_Post_Process()
...
+    Coord3D c2;
```
Comment (xezon) [inline src/game/logic/map/terrainlogic.cpp]: rotated_axis

File: src/game/logic/map/terrainlogic.cpp [hunk 15]
```diff
@@ -1416,3 +1446,845 @@ void TerrainLogic::Load_Post_Process()
...
+
+    if (c.z != 0.0f) {
+        c2.z = -(c2.x * c.x + c2.y * c.y) / c.z;
+        c2.Normalize();
+    }
+
+    captainslog_dbgassert(GameMath::Fabs(c2.x * c.x + c2.y * c.y + c2.z * c.z) < 0.0001f,
+        "dot is not zero (%f)",
+        GameMath::Fabs(c2.x * c.x + c2.y * c.y + c2.z * c.z));
+    Coord3D c3;
```
Comment (xezon) [inline src/game/logic/map/terrainlogic.cpp]: perpendicular_axis

File: src/game/logic/map/terrainlogic.cpp [hunk 16]
```diff
@@ -1416,3 +1446,845 @@ void TerrainLogic::Load_Post_Process()
...
+        region->lo.y = extent.hi.y + 99999.898f;
+        region->hi.x = extent.lo.x - 99999.898f;
+        region->hi.y = extent.lo.y - 99999.898f;
+
+        if (water == &m_gridWaterHandle) {
+            float grid_cells_x;
+            float grid_cells_y;
+            float cell_size;
+            g_theTerrainVisual->Get_Water_Grid_Resolution(water, &grid_cells_x, &grid_cells_y, &cell_size);
+            ICoord3D p[4];
```
Comment (xezon) [inline src/game/logic/map/terrainlogic.cpp]: points

File: src/game/logic/map/terrainlogic.cpp [hunk 17]
```diff
@@ -1416,3 +1446,845 @@ void TerrainLogic::Load_Post_Process()
...
+            p[0].x = 0;
+            p[0].y = 0;
+            p[1].x = grid_cells_x * cell_size;
+            p[1].y = 0;
+            p[2].x = p[1].x;
+            p[2].y = grid_cells_y * cell_size;
+            p[3].x = 0;
+            p[3].y = p[2].y;
+            Vector3 v;
```
Comment (xezon) [inline src/game/logic/map/terrainlogic.cpp]: tp for transformed point.

File: src/game/logic/map/terrainlogic.cpp [hunk 18]
```diff
@@ -1416,3 +1446,845 @@ void TerrainLogic::Load_Post_Process()
...
+    }
+
+    return nullptr;
+}
+
+void TerrainLogic::Flatten_Terrain(Object *obj)
+{
+    if (!obj->Get_Geometry_Info().Is_Small()) {
+        const Coord3D *pos = obj->Get_Position();
+        GeometryType type = obj->Get_Geometry_Info().Get_Type();
```
Comment (xezon) [inline src/game/logic/map/terrainlogic.cpp]: geometry_type

File: src/game/logic/map/terrainlogic.cpp [hunk 19]
```diff
@@ -1416,3 +1446,845 @@ void TerrainLogic::Load_Post_Process()
...
+
+        if (type >= GEOMETRY_SPHERE) {
+            if (type <= GEOMETRY_CYLINDER) {
+                float radius = obj->Get_Geometry_Info().Get_Major_Radius();
```
Comment (xezon) [inline src/game/logic/map/terrainlogic.cpp]: major_radius

File: src/game/logic/map/terrainlogic.cpp [hunk 20]
```diff
@@ -1416,3 +1446,845 @@ void TerrainLogic::Load_Post_Process()
...
+                float radius_square = GameMath::Square(radius);
+                int x_min = GameMath::Fast_To_Int_Floor((pos->x - radius) / 10.0f);
+                int y_min = GameMath::Fast_To_Int_Floor((pos->y - radius) / 10.0f);
+                int x_max = GameMath::Fast_To_Int_Floor((pos->x + radius) / 10.0f);
+                int y_max = GameMath::Fast_To_Int_Floor((pos->y + radius) / 10.0f);
+                float f1 = 0.0f;
```
Comment (xezon) [inline src/game/logic/map/terrainlogic.cpp]: ground_height_sum

File: src/game/logic/map/terrainlogic.cpp [hunk 21]
```diff
@@ -1416,3 +1446,845 @@ void TerrainLogic::Load_Post_Process()
...
+
+                        if (b) {
+                            f1 += g_theTerrainLogic->Get_Ground_Height(v.X, v.Y, nullptr);
+                            count++;
+                        }
+                    }
+                }
+
+                if (count != 0) {
+                    int height = GameMath::Fast_To_Int_Floor((f1 / count) / HEIGHTMAP_SCALE + 0.5f);
```
Comment (xezon) [inline src/game/logic/map/terrainlogic.cpp]: flattened_ground_height

File: src/game/logic/map/terrainlogic.cpp [hunk 22]
```diff
@@ -1416,3 +1446,845 @@ void TerrainLogic::Load_Post_Process()
...
+                float radius = obj->Get_Geometry_Info().Get_Major_Radius();
+                float radius_square = GameMath::Square(radius);
+                int x_min = GameMath::Fast_To_Int_Floor((pos->x - radius) / 10.0f);
+                int y_min = GameMath::Fast_To_Int_Floor((pos->y - radius) / 10.0f);
+                int x_max = GameMath::Fast_To_Int_Floor((pos->x + radius) / 10.0f);
+                int y_max = GameMath::Fast_To_Int_Floor((pos->y + radius) / 10.0f);
+                float f1 = 0.0f;
+                int count = 0;
+
+                for (int i = x_min; i <= x_max; i++) {
```
Comment (xezon) [inline src/game/logic/map/terrainlogic.cpp]: x

File: src/game/logic/map/terrainlogic.cpp [hunk 23]
```diff
@@ -1416,3 +1446,845 @@ void TerrainLogic::Load_Post_Process()
...
+                    for (int j = 0; j <= y_max; j++) {
```
Comment (xezon) [inline src/game/logic/map/terrainlogic.cpp]: y

File: src/game/logic/map/terrainlogic.cpp [hunk 24]
```diff
@@ -1416,3 +1446,845 @@ void TerrainLogic::Load_Post_Process()
...
+        if (type >= GEOMETRY_SPHERE) {
+            if (type <= GEOMETRY_CYLINDER) {
+                float radius = obj->Get_Geometry_Info().Get_Major_Radius();
+                float radius_square = GameMath::Square(radius);
+                int x_min = GameMath::Fast_To_Int_Floor((pos->x - radius) / 10.0f);
+                int y_min = GameMath::Fast_To_Int_Floor((pos->y - radius) / 10.0f);
+                int x_max = GameMath::Fast_To_Int_Floor((pos->x + radius) / 10.0f);
+                int y_max = GameMath::Fast_To_Int_Floor((pos->y + radius) / 10.0f);
+                float f1 = 0.0f;
+                int count = 0;
```
Comment (xezon) [inline src/game/logic/map/terrainlogic.cpp]: ground_height_sample_count

File: src/game/logic/map/terrainlogic.cpp [hunk 25]
```diff
@@ -1416,3 +1446,845 @@ void TerrainLogic::Load_Post_Process()
...
+                            f1 += g_theTerrainLogic->Get_Ground_Height(v.X, v.Y, nullptr);
+                            count++;
+                        }
+                    }
+                }
+
+                if (count != 0) {
+                    int height = GameMath::Fast_To_Int_Floor((f1 / count) / HEIGHTMAP_SCALE + 0.5f);
+
+                    for (int i = x_min; i <= x_max; i++) {
```
Comment (xezon) [inline src/game/logic/map/terrainlogic.cpp]: x

File: src/game/logic/map/terrainlogic.cpp [hunk 26]
```diff
@@ -1416,3 +1446,845 @@ void TerrainLogic::Load_Post_Process()
...
+                        for (int j = 0; j <= y_max; j++) {
```
Comment (xezon) [inline src/game/logic/map/terrainlogic.cpp]: y

File: src/game/logic/map/terrainlogic.cpp [hunk 27]
```diff
@@ -1416,3 +1446,845 @@ void TerrainLogic::Load_Post_Process()
...
+
+                if (max_y < v4.Y) {
+                    max_y = v4.Y;
+                }
+
+                int x_min = GameMath::Fast_To_Int_Floor(min_x / 10.0f);
+                int y_min = GameMath::Fast_To_Int_Floor(min_y / 10.0f);
+                int x_max = GameMath::Fast_To_Int_Floor(max_x / 10.0f);
+                int y_max = GameMath::Fast_To_Int_Floor(max_y / 10.0f);
+                float f1 = 0.0f;
```
Comment (xezon) [inline src/game/logic/map/terrainlogic.cpp]: ground_height_sum

File: src/game/logic/map/terrainlogic.cpp [hunk 28]
```diff
@@ -1416,3 +1446,845 @@ void TerrainLogic::Load_Post_Process()
...
+                int count = 0;
+
+                for (int i = x_min; i <= x_max; i++) {
```
Comment (xezon) [inline src/game/logic/map/terrainlogic.cpp]: x

File: src/game/logic/map/terrainlogic.cpp [hunk 29]
```diff
@@ -1416,3 +1446,845 @@ void TerrainLogic::Load_Post_Process()
...
+                    for (int j = 0; j <= y_max; j++) {
```
Comment (xezon) [inline src/game/logic/map/terrainlogic.cpp]: y

File: src/game/logic/map/terrainlogic.cpp [hunk 30]
```diff
@@ -1416,3 +1446,845 @@ void TerrainLogic::Load_Post_Process()
...
+                if (max_y < v4.Y) {
+                    max_y = v4.Y;
+                }
+
+                int x_min = GameMath::Fast_To_Int_Floor(min_x / 10.0f);
+                int y_min = GameMath::Fast_To_Int_Floor(min_y / 10.0f);
+                int x_max = GameMath::Fast_To_Int_Floor(max_x / 10.0f);
+                int y_max = GameMath::Fast_To_Int_Floor(max_y / 10.0f);
+                float f1 = 0.0f;
+                int count = 0;
```
Comment (xezon) [inline src/game/logic/map/terrainlogic.cpp]: ground_height_sample_count

File: src/game/logic/map/terrainlogic.cpp [hunk 31]
```diff
@@ -1416,3 +1446,845 @@ void TerrainLogic::Load_Post_Process()
...
+
+                        if (b) {
+                            f1 += g_theTerrainLogic->Get_Ground_Height(v.X, v.Y, nullptr);
+                            count++;
+                        }
+                    }
+                }
+
+                if (count != 0) {
+                    int height = GameMath::Fast_To_Int_Floor((f1 / count) / HEIGHTMAP_SCALE + 0.5f);
```
Comment (xezon) [inline src/game/logic/map/terrainlogic.cpp]: flattened_ground_height

File: src/game/logic/map/terrainlogic.cpp [hunk 32]
```diff
@@ -1416,3 +1446,845 @@ void TerrainLogic::Load_Post_Process()
...
+                    int height2 = GameMath::Fast_To_Int_Floor(
+                        g_theTerrainLogic->Get_Ground_Height(pos->x, pos->y, nullptr) / HEIGHTMAP_SCALE);
+
+                    if (height > height2) {
+                        height = height2;
+                    }
+
+                    for (int i = x_min; i <= x_max; i++) {
```
Comment (xezon) [inline src/game/logic/map/terrainlogic.cpp]: x

File: src/game/logic/map/terrainlogic.cpp [hunk 33]
```diff
@@ -1416,3 +1446,845 @@ void TerrainLogic::Load_Post_Process()
...
+                        for (int j = 0; j <= y_max; j++) {
```
Comment (xezon) [inline src/game/logic/map/terrainlogic.cpp]: y

File: src/game/logic/map/terrainlogic.cpp [hunk 34]
```diff
@@ -1416,3 +1446,845 @@ void TerrainLogic::Load_Post_Process()
...
+                int x_max = GameMath::Fast_To_Int_Floor(max_x / 10.0f);
+                int y_max = GameMath::Fast_To_Int_Floor(max_y / 10.0f);
+                float f1 = 0.0f;
+                int count = 0;
+
+                for (int i = x_min; i <= x_max; i++) {
+                    for (int j = 0; j <= y_max; j++) {
+                        Vector3 v(i * 10.0f, j * 10.0f, 0.0f);
+                        unsigned char flags;
+                        bool b = Point_In_Triangle_2D(v1, v2, v4, v, 0, 1, flags);
```
Comment (xezon) [inline src/game/logic/map/terrainlogic.cpp]: inside

File: src/game/logic/map/terrainlogic.cpp [hunk 35]
```diff
@@ -1416,3 +1446,845 @@ void TerrainLogic::Load_Post_Process()
...
+
+                    if (height > height2) {
+                        height = height2;
+                    }
+
+                    for (int i = x_min; i <= x_max; i++) {
+                        for (int j = 0; j <= y_max; j++) {
+                            Vector3 v(j * 10.0f, i * 10.0f, 0.0f);
+                            unsigned char flags;
+                            bool b = Point_In_Triangle_2D(v1, v2, v4, v, 0, 1, flags);
```
Comment (xezon) [inline src/game/logic/map/terrainlogic.cpp]: inside

File: src/game/logic/map/terrainlogic.cpp [hunk 36]
```diff
@@ -1416,3 +1446,845 @@ void TerrainLogic::Load_Post_Process()
...
+                    }
+                }
+
+                if (count != 0) {
+                    int height = GameMath::Fast_To_Int_Floor((f1 / count) / HEIGHTMAP_SCALE + 0.5f);
+
+                    for (int i = x_min; i <= x_max; i++) {
+                        for (int j = 0; j <= y_max; j++) {
+                            Vector3 v(i * 10.0f, j * 10.0f, 0.0f);
+                            bool b = false;
```
Comment (xezon) [inline src/game/logic/map/terrainlogic.cpp]: inside

File: src/game/logic/map/terrainlogic.cpp [hunk 37]
```diff
@@ -1416,3 +1446,845 @@ void TerrainLogic::Load_Post_Process()
...
+                int y_min = GameMath::Fast_To_Int_Floor((pos->y - radius) / 10.0f);
+                int x_max = GameMath::Fast_To_Int_Floor((pos->x + radius) / 10.0f);
+                int y_max = GameMath::Fast_To_Int_Floor((pos->y + radius) / 10.0f);
+                float f1 = 0.0f;
+                int count = 0;
+
+                for (int i = x_min; i <= x_max; i++) {
+                    for (int j = 0; j <= y_max; j++) {
+                        Vector3 v(i * 10.0f, j * 10.0f, 0.0f);
+                        bool b = false;
```
Comment (xezon) [inline src/game/logic/map/terrainlogic.cpp]: inside

File: src/game/logic/map/terrainlogic.cpp [hunk 38]
```diff
@@ -1416,3 +1446,845 @@ void TerrainLogic::Load_Post_Process()
...
+            }
+        }
+    }
+}
+
+void TerrainLogic::Create_Crater_In_Terrain(Object *obj)
+{
+    if (!obj->Get_Geometry_Info().Is_Small()) {
+        const Coord3D *pos = obj->Get_Position();
+        float radius = obj->Get_Geometry_Info().Get_Major_Radius();
```
Comment (xezon) [inline src/game/logic/map/terrainlogic.cpp]: major_radius

File: src/game/logic/map/terrainlogic.cpp [hunk 39]
```diff
@@ -1416,3 +1446,845 @@ void TerrainLogic::Load_Post_Process()
...
+
+        if (radius > 0.0f) {
+            int x_min = GameMath::Fast_To_Int_Floor((pos->x - radius) / 10.0f);
+            int y_min = GameMath::Fast_To_Int_Floor((pos->y - radius) / 10.0f);
+            int x_max = GameMath::Fast_To_Int_Floor((pos->x + radius) / 10.0f);
+            int y_max = GameMath::Fast_To_Int_Floor((pos->y + radius) / 10.0f);
+
+            for (int i = x_min; i <= x_max; i++) {
```
Comment (xezon) [inline src/game/logic/map/terrainlogic.cpp]: x

File: src/game/logic/map/terrainlogic.cpp [hunk 40]
```diff
@@ -1416,3 +1446,845 @@ void TerrainLogic::Load_Post_Process()
...
+                for (int j = 0; j <= y_max; j++) {
```
Comment (xezon) [inline src/game/logic/map/terrainlogic.cpp]: y

File: src/game/logic/map/terrainlogic.cpp [hunk 41]
```diff
@@ -1416,3 +1446,845 @@ void TerrainLogic::Load_Post_Process()
...
+                    float f = GameMath::Sqrt(GameMath::Square(i * 10.0f) + GameMath::Square(j * 10.0f));
```
Comment (xezon) [inline src/game/logic/map/terrainlogic.cpp]: crater_radius

File: src/game/logic/map/terrainlogic.cpp [hunk 42]
```diff
@@ -1416,3 +1446,845 @@ void TerrainLogic::Load_Post_Process()
...
+
+                    if (f < radius) {
+                        ICoord2D c;
+                        c.x = i;
+                        c.y = j;
+                        float f2 = (1.0f - f / radius) * radius;
```
Comment (xezon) [inline src/game/logic/map/terrainlogic.cpp]: crater_depth

File: src/game/logic/map/terrainlogic.cpp [hunk 43]
```diff
@@ -1416,3 +1446,845 @@ void TerrainLogic::Load_Post_Process()
...
+        }
+    }
+}
+
+void TerrainLogic::Update()
+{
+    m_bridgeDamageStatesChanged = false;
+
+    if (m_numWaterToUpdate != 0) {
+        bool b = g_theGameLogic->Get_Frame() % 30 == 0;
```
Comment (xezon) [inline src/game/logic/map/terrainlogic.cpp]: deal_damage

File: src/game/logic/map/terrainlogic.cpp [hunk 44]
```diff
@@ -1416,3 +1446,845 @@ void TerrainLogic::Load_Post_Process()
...
+
+        for (int i = m_numWaterToUpdate - 1; i >= 0; i--) {
+            const WaterHandle *water_table = m_waterToUpdate[i].water_table;
+            float change_per_frame = m_waterToUpdate[i].change_per_frame;
+            float target_height = m_waterToUpdate[i].target_height;
+            float damage_amount = m_waterToUpdate[i].damage_amount;
+            float current_height = m_waterToUpdate[i].current_height;
+
+            bool b2 = false;
```
Comment (xezon) [inline src/game/logic/map/terrainlogic.cpp]: update_pathfind

File: src/platform/w3dengine/client/w3dterrainlogic.cpp (hunk 1 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3dterrainlogic.cpp]: fixed.

File: src/platform/w3dengine/client/w3dterrainlogic.cpp (hunk 2 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3dterrainlogic.cpp]: fixed.

File: src/platform/w3dengine/client/w3dterrainlogic.cpp (hunk 4 above)
Comment (jonwil) [reply src/platform/w3dengine/client/w3dterrainlogic.cpp:198]: This is what the original is doing.

File: src/game/logic/map/terrainlogic.cpp (hunk 5 above)
Comment (jonwil) [reply src/game/logic/map/terrainlogic.cpp]: fixed.

File: src/game/logic/map/terrainlogic.cpp (hunk 8 above)
Comment (jonwil) [reply src/game/logic/map/terrainlogic.cpp]: fixed.

File: src/game/logic/map/terrainlogic.cpp (hunk 10 above)
Comment (jonwil) [reply src/game/logic/map/terrainlogic.cpp]: fixed.

File: src/game/logic/map/terrainlogic.cpp (hunk 9 above)
Comment (jonwil) [reply src/game/logic/map/terrainlogic.cpp]: fixed.

File: src/game/logic/map/terrainlogic.cpp (hunk 11 above)
Comment (jonwil) [reply src/game/logic/map/terrainlogic.cpp]: fixed.

File: src/game/logic/map/terrainlogic.cpp (hunk 12 above)
Comment (jonwil) [reply src/game/logic/map/terrainlogic.cpp]: fixed.

File: src/game/logic/map/terrainlogic.cpp (hunk 13 above)
Comment (jonwil) [reply src/game/logic/map/terrainlogic.cpp]: fixed.

File: src/game/logic/map/terrainlogic.cpp (hunk 14 above)
Comment (jonwil) [reply src/game/logic/map/terrainlogic.cpp]: fixed.

File: src/game/logic/map/terrainlogic.cpp (hunk 15 above)
Comment (jonwil) [reply src/game/logic/map/terrainlogic.cpp]: fixed.

File: src/game/logic/map/terrainlogic.cpp (hunk 17 above)
Comment (jonwil) [reply src/game/logic/map/terrainlogic.cpp]: fixed.

File: src/game/logic/map/terrainlogic.cpp (hunk 16 above)
Comment (jonwil) [reply src/game/logic/map/terrainlogic.cpp]: fixed.

File: src/game/logic/map/terrainlogic.cpp (hunk 18 above)
Comment (jonwil) [reply src/game/logic/map/terrainlogic.cpp]: fixed.

File: src/game/logic/map/terrainlogic.cpp (hunk 19 above)
Comment (jonwil) [reply src/game/logic/map/terrainlogic.cpp]: fixed.

File: src/game/logic/map/terrainlogic.cpp (hunk 20 above)
Comment (jonwil) [reply src/game/logic/map/terrainlogic.cpp]: fixed.

File: src/game/logic/map/terrainlogic.cpp (hunk 21 above)
Comment (jonwil) [reply src/game/logic/map/terrainlogic.cpp]: fixed.

File: src/game/logic/map/terrainlogic.cpp (hunk 22 above)
Comment (jonwil) [reply src/game/logic/map/terrainlogic.cpp]: fixed.

File: src/game/logic/map/terrainlogic.cpp (hunk 23 above)
Comment (jonwil) [reply src/game/logic/map/terrainlogic.cpp]: fixed.

File: src/game/logic/map/terrainlogic.cpp (hunk 24 above)
Comment (jonwil) [reply src/game/logic/map/terrainlogic.cpp]: fixed.

File: src/game/logic/map/terrainlogic.cpp (hunk 25 above)
Comment (jonwil) [reply src/game/logic/map/terrainlogic.cpp]: fixed.

File: src/game/logic/map/terrainlogic.cpp (hunk 26 above)
Comment (jonwil) [reply src/game/logic/map/terrainlogic.cpp]: fixed.

File: src/game/logic/map/terrainlogic.cpp (hunk 27 above)
Comment (jonwil) [reply src/game/logic/map/terrainlogic.cpp]: fixed.

File: src/game/logic/map/terrainlogic.cpp (hunk 28 above)
Comment (jonwil) [reply src/game/logic/map/terrainlogic.cpp]: fixed.

File: src/game/logic/map/terrainlogic.cpp (hunk 29 above)
Comment (jonwil) [reply src/game/logic/map/terrainlogic.cpp]: fixed.

File: src/game/logic/map/terrainlogic.cpp (hunk 30 above)
Comment (jonwil) [reply src/game/logic/map/terrainlogic.cpp]: fixed.

File: src/game/logic/map/terrainlogic.cpp (hunk 31 above)
Comment (jonwil) [reply src/game/logic/map/terrainlogic.cpp]: fixed.

File: src/game/logic/map/terrainlogic.cpp (hunk 32 above)
Comment (jonwil) [reply src/game/logic/map/terrainlogic.cpp]: fixed.

File: src/game/logic/map/terrainlogic.cpp (hunk 33 above)
Comment (jonwil) [reply src/game/logic/map/terrainlogic.cpp]: fixed.

File: src/game/logic/map/terrainlogic.cpp (hunk 34 above)
Comment (jonwil) [reply src/game/logic/map/terrainlogic.cpp]: fixed.

File: src/game/logic/map/terrainlogic.cpp (hunk 35 above)
Comment (jonwil) [reply src/game/logic/map/terrainlogic.cpp]: fixed.

File: src/game/logic/map/terrainlogic.cpp (hunk 36 above)
Comment (jonwil) [reply src/game/logic/map/terrainlogic.cpp]: fixed.

File: src/game/logic/map/terrainlogic.cpp (hunk 37 above)
Comment (jonwil) [reply src/game/logic/map/terrainlogic.cpp]: fixed.

File: src/game/logic/map/terrainlogic.cpp (hunk 38 above)
Comment (jonwil) [reply src/game/logic/map/terrainlogic.cpp]: fixed.

File: src/game/logic/map/terrainlogic.cpp (hunk 39 above)
Comment (jonwil) [reply src/game/logic/map/terrainlogic.cpp]: fixed.

File: src/game/logic/map/terrainlogic.cpp (hunk 40 above)
Comment (jonwil) [reply src/game/logic/map/terrainlogic.cpp]: fixed.

File: src/game/logic/map/terrainlogic.cpp (hunk 41 above)
Comment (jonwil) [reply src/game/logic/map/terrainlogic.cpp]: fixed.

File: src/game/logic/map/terrainlogic.cpp (hunk 42 above)
Comment (jonwil) [reply src/game/logic/map/terrainlogic.cpp]: fixed.

File: src/game/logic/map/terrainlogic.cpp (hunk 43 above)
Comment (jonwil) [reply src/game/logic/map/terrainlogic.cpp]: fixed.

File: src/game/logic/map/terrainlogic.cpp (hunk 44 above)
Comment (jonwil) [reply src/game/logic/map/terrainlogic.cpp]: fixed.

**Comment by xezon** [comment]
Shell map looks as usual.

Flattening Terrain with structure placements does not work. Tested on Alpine Assault (2).

shot_20230422_222952_1

**Comment by xezon** [comment]
There appear to be discrepancies with

1. shape of water at shore looks different
~~2. map boundaries - can see less of the border in Thyme~~

Tested on

[[RANK] Melting Snow ZH v4.zip](https://github.com/TheAssemblyArmada/Thyme/files/11302256/RANK.Melting.Snow.ZH.v4.zip)


## Original

shot_20230422_224138_1

## Thyme

shot_20230422_224239_1

**Comment by jonwil** [comment]
Tested with thyme.dll from build artifacts https://github.com/TheAssemblyArmada/Thyme/actions/runs/4709737215 and this is what I get:
water
Confirms that the water issue is not caused by this commit.
No clue what I am supposed to be looking at for the border problem, I can't tell what the difference is.
Confirmed the flattening issue, likely a bug in TerrainLogic::Flatten_Terrain

**Comment by jonwil** [comment]
Did a thorough re-do of TerrainLogic::Flatten_Terrain and found one issue that I fixed but I can't see any other issues with the function or anything else that could possibly be causing the terrain flattening issue and I have no idea how to further debug or fix it at this point (assuming it's still a thing after my fix)

**Comment by xezon** [comment]
Shoreline issue tracked with #902

**Review by xezon — APPROVED** [review]
Maps load
Flattening terrain works
Rising water works
Units taking damage in water works
Angling units to terrain works
Cliff climbing works

Looks okay as far as I can tell.

**Pull request merged**